etc/delphix-platform/ansible/apply

```diff
@@ -42,6 +42,42 @@ function apply_playbook() {
 	fi
 }
 
+#
+# To reduce the bootup latency and the latency to start services that
+# depend on this service, we only want to apply the configuration a
+# single time. This way, the configuration will be applied on the first
+# boot, but not on subsequent boots. This will allow the bootup time for
+# the subsequent boots to be quicker.
+#
+# Thus, if this "ansible-done" file is present, this signifies that the
+# ansible configuration has already been applied, so there's no need to
+# apply this configuration again.
+#
+# We don't use Systemd's "ConditionPathExists" directive to disable the
+# service after it's been run once, for a couple of reasons:
+#
+# 1. That would cause the service to enter a "failed" state when the
+#    file exists. This isn't the behavior we want, since it then makes
+#    it more difficult for other services to depend on this service
+#    running successfully.
+#
+# 2. When the file exists, restarting "delphix-platform" would not
+#    result in restarting services that depend on "delphix-platform". We
+#    rely on restarts to "delphix-platfrom" causing restarts to all
+#    dependent services, so we need to ensure this functionality
+#    continues to work, even if the ansible configuration doesn't have
+#    to be re-applied.
+#
+# Also worth noting, consumers of this service (e.g. upgrade, factory
+# reset, etc.) are aware of this file and it's implications. Thus, if we
+# ever need to change this file's path, we'll likely have to also update
+# all of the consumers of this service; since these consumers may rely
+# on removing this file, in order to force this service to re-run.
+#
+if [[ -f /etc/delphix-platform/ansible-done ]]; then
```

Comment:
I think this would be right right place to check for DLPX_FACTORY_RESET and remove the ansible-done file if the variable is true.

Reply:
I disagree; the `DLPX_FACTORY_RESET` variable is an app-gate concept, so leaking that into this repository doesn't seem like a good design. Currently, this repository knows nothing about that variable.

Further, I think we need a way to re-run the ansible configuration, without invoking the app-gate's factory reset logic. E.g. on upgrade, we want to run the ansible code without invoking factory reset. If we only had a single tunable, `DLPX_FACTORY_RESET`, I'm not sure how we could invoke the logic unrelated to factory reset, without also invoking the factory reset logic (and breaking the customer's engine).

What do you think?

Reply:
I understand your point of view.

My perspective on factory reset when making the suggestion was that I viewed setting DLPX_FACTORY_RESET on delphix-platform as the only way to perform a factory reset of the entire appliance today. The variable is exposed by the delphix-platform service insofar as there is no other way to do a factory reset.

I think your point is that this design isn't a good one, and we should separately re-design factory reset so that it doesn't use this environment variable, and also doesn't rely on being driven by packages' ansible roles. Is that right? I'm fine with that, let's create a jira issue for that.

Reply:
My point is this, delphix-platform doesn't have a concept of "factory reset". This concept is created by the app-gate, and doesn't exist outside of the app-gate. Any appliance we create that doesn't have the app-gate on it (e.g. dc1), has no concept of "factory reset", even if it has the delphix-platform installed.

Further, the way the app-gate implements factory reset on Linux is not very elegant, or extensible (as you've mentioned). Thus, I'd like to ignore how it's currently implemented (i.e. forget about the specific variable) for now, and think more broadly w.r.t. the concepts we're trying to accomplish.

If we want to extend delphix-platform to have a concept of "factory reset" (which is what you're suggesting, by checking for the `DLPX_FACTORY_RESET` variable here), such that that other consumers can make use of it (e.g. app-gate uses this new concept to implement it's own factory reset logic, masking uses it do the same, etc.), I'm OK with that. But, it's more work that would need to be scoped and implemented (since I'm not sure just porting the current implementation is the best way to do it).

Additionally, in my mind, that functionality is orthogonal to implementing this "configure once" concept. We could implement "configure once" without implementing "factory reset" in delphix-platform (which I'm trying to do in this PR), and could also implement "factory reset" without implementing "configure once".

What I think you're suggesting is that we implement both "configure once" and "factory reset" in delphix-platform, because once the two are implemented in delphix-plaform, consumers of delphix-platform become more simple (e.g. the app-gate logic you commented on could be more simple).

I'm OK going that route, it's just more work (both design and implementation) than I was expecting to tackle with this PR. So, if we want to implement both concepts simultaneously in delphix-platform because ultimately that'll lead to a better overall solution, I'm OK with that, but it'll have to be something that is left to be implemented later (if at all).

I see two ways to proceed:

1. Close this PR and wait for "configure once" and "factory reset" to be implemented in delphix-platform at the same time.

2. Move forward with this PR, and only tackle the "configure once" semantics which I was going for; potentially opening a JIRA issue to tackle "factory reset" in delphix-platform later. As pointed out in the app-gate review, this might leave is with a sub-optimal solution given the current "factory reset" implementation.

I'm happy to go with either approach, honestly. As long as we're all on the same page as to which we'd like to go with. Which would you prefer?

What I don't want to do is implement a poor solution for factory reset in delphix-platform, which is then difficult to modify and improve later, because it's consumed by multiple different projects (and would require a multi-repository and/or multi-team flag day to reconcile). I think we might end up in this situation if we just simply checked for `DLPX_FACTORY_RESET` here.

Reply:
Yep, I think we can move on, we also hashed this out in the dlpx-app-gate review. I'm fine with the change as-is and our handling factory reset changes separately.